app/frontend/wallet/shelley/transaction/computeTxPlan.ts

```diff
@@ -0,0 +1,327 @@
+    [...outputs, ...tokenChangeOutputs],
+    certificates,
+    withdrawals,
+    auxiliaryData
+  )
+
+  const minimalTokenChangeLovelace = tokenChangeOutputs.reduce(
+    (acc, {coins}) => acc + coins,
+    0
+  ) as Lovelace
+
+  const remainingTokenChangeLovelace = (totalInput -
+    totalOutput -
+    minimalTokenChangeLovelace -
+    feeWithTokenChange) as Lovelace
+
+  // remainingTokenChangeLovelace has to be positive,
+  // otherwise not enough funds to pay for change outputs
+  if (remainingTokenChangeLovelace < 0) {
+    return {
+      success: false,
+      minimalLovelaceAmount: additionalLovelaceAmount,
+      estimatedFee: feeWithTokenChange,
+      deposit,
+      error: {code: UnexpectedErrorReason.CannotConstructTxPlan},
+    }
+  }
+
+  // if remainingTokenChangeLovelace is positive, we try to put it in ada only change output
+  if (remainingTokenChangeLovelace > MIN_UTXO_VALUE) {
+    const feeWithAdaAndTokenChange = computeRequiredTxFee(
+      inputs,
+      [...outputs, ...tokenChangeOutputs, adaOnlyChangeOutput],
+      certificates,
+      withdrawals,
+      auxiliaryData
+    )
+
+    const adaOnlyChangeOutputLovelace = (remainingTokenChangeLovelace -
+      (feeWithAdaAndTokenChange - feeWithTokenChange)) as Lovelace
+
+    return {
+      success: true,
+      txPlan: {
+        inputs,
+        outputs,
+        change: [
+          {...adaOnlyChangeOutput, coins: adaOnlyChangeOutputLovelace},
+          ...tokenChangeOutputs,
+        ],
+        certificates,
+        deposit,
+        additionalLovelaceAmount,
+        fee: feeWithAdaAndTokenChange,
+        baseFee: feeWithAdaAndTokenChange,
+        withdrawals,
+        auxiliaryData,
+      },
+    }
+  }
+
+  // if remainingTokenChangeLovelace is too small for separate output,
+  // we add the remainingTokenChangeLovelace to the first changeOutput
+
+  const firstChangeOutput: TxOutput = {
+    ...tokenChangeOutputs[0],
+    coins: (tokenChangeOutputs[0].coins + remainingTokenChangeLovelace) as Lovelace,
+  }
+
+  return {
+    success: true,
+    txPlan: {
+      inputs,
+      outputs,
+      change: tokenChangeOutputs.map((output, i) => (i === 0 ? firstChangeOutput : output)),
+      certificates,
+      deposit,
+      additionalLovelaceAmount,
+      fee: feeWithTokenChange,
+      baseFee: feeWithTokenChange,
+      withdrawals,
+      auxiliaryData,
+    },
+  }
+}
+
+export const validateTxPlan = (txPlanResult: TxPlanResult): TxPlanResult => {
+  if (txPlanResult.success === false) {
+    return txPlanResult
+  }
+  const {txPlan} = txPlanResult
+  const {
+    change,
+    outputs,
+    withdrawals,
+    fee,
+    additionalLovelaceAmount,
+    certificates,
+    deposit,
+    baseFee,
+  } = txPlan
+
+  const noTxPlan: TxPlanResult = {
+    success: false,
+    error: null,
+    estimatedFee: fee,
+    deposit,
+    minimalLovelaceAmount: additionalLovelaceAmount,
+  }
+
+  const outputsWithChange = change ? [...outputs, ...change] : outputs
```

Comment:
empty array is truthy in JS, so I believe this actually always ends up in the left branch, or can the "change" be null? In that case I'd rathe rbe more explicit and check for `change != null`

Reply:
I totally missed this, thanks, will fix :+1: